### PR TITLE
token-upgrade: Add CLI and docs

### PR DESCRIPTION
**Pull request by joncinque**
#### Problem

The token-upgrade program will be added in #3436, but there's no usable client or usage information.

#### Solution

Building on #3436, add a CLI based on https://github.com/mvines/solana-cli-template along with documentation of how to use the program.

### CODE REVIEW

**Comment by 2501babe** [comment]
this looks good, based on a rough first read. going to go over it again in more detail soon

File: docs/src/token-upgrade.md [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+delegated to this escrow authority may be used in the protocol.
+
+A holder of original tokens provides their original token account, a new token account,
+and a new token account owned by or delegated to the escrow authority. If the
+escrow authority's account has enough tokens, the protocol will burn the original
+tokens and transfer the same amount of new tokens to the user's new account.
+
+The amount of tokens transferred is completely independent of decimals, so if
+the mints have different decimals, then users may be surprised to see a different
+amount.
```
Comment (2501babe) [inline docs/src/token-upgrade.md:31]: this should be updated

**Review by 2501babe — DISMISSED** [review]

File: docs/src/token-upgrade.md [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+Token-2022 contains many new features for mint owners to customize the behavior
+of their tokens. You can find full information about Token-2022 and its extensions
+in the [documentation](token-2022.md).
+
+Mint owners may want to take advantage of new functionality for their users, but
+there is no way to automatically convert tokens from Token to Token-2022.
+
+The Token Upgrade Program defines an escrow authority, a program-derived address,
+from two addresses, the original and new mints. Any new token account owned by or
```
Comment (CriesofCarrots) [inline docs/src/token-upgrade.md:21]:
```suggestion
The Token Upgrade Program defines an escrow authority, a program-derived address
from two addresses, the original and new mints. Any new token account owned by or
```

File: docs/src/token-upgrade.md [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+delegated to this escrow authority may be used in the protocol.
```
Comment (CriesofCarrots) [inline docs/src/token-upgrade.md:22]:
```suggestion
delegated to this escrow authority may be used as the escrow account.
```

File: docs/src/token-upgrade.md [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+A holder of original tokens provides their original token account, a new token account,
+and a new token account owned by or delegated to the escrow authority. If the
+escrow authority's account has enough tokens, the protocol will burn the original
```
Comment (CriesofCarrots) [inline docs/src/token-upgrade.md:26]:
```suggestion
A holder of original tokens provides their original token account, a new token account,
and the escrow account. If the
escrow account has enough tokens, the protocol will burn the original
```

File: docs/src/token-upgrade.md [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+tokens and transfer the same amount of new tokens to the user's new account.
+
+The amount of tokens transferred is completely independent of decimals, so if
+the mints have different decimals, then users may be surprised to see a different
+amount.
+
+The program is completely stateless and has a simple implementation, so mint owners
+may customize it if they need additional functionality. For example, if they want
+the transfer to be decimal-aware, they can scale the transferred number up or down
+as needed.
```
Comment (CriesofCarrots) [inline docs/src/token-upgrade.md:36]: This reference to decimals also needs to be updated, right?

**Review by CriesofCarrots — COMMENTED** [review]
A couple small docs suggestions. Haven't looked at main.rs yet; coming back to it in a bit.

File: docs/src/token-upgrade.md (hunk 1 above)
Comment (joncinque) [reply docs/src/token-upgrade.md:31]: Thanks for spotting this

File: token-upgrade/cli/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,831 @@
...
+                    .takes_value(true)
+                    .help("Specify the burnt account address. [default: associated token account for owner on original mint]"),
+            )
+            .arg(
+                Arg::new("escrow")
+                    .long("escrow")
+                    .value_name("ESCROW_TOKEN_ACCOUNT_ADDRESS")
+                    .validator(|s| is_valid_pubkey(s))
+                    .takes_value(true)
+                    .help("Specify the escrow account address to transfer from. [default: associated token account for the escrow authority]"),
```
Comment (CriesofCarrots) [inline token-upgrade/cli/src/main.rs:317]:
```suggestion
                    .help("Specify the escrow account address to transfer from. [default: associated token account for the escrow authority on new mint]"),
```

File: token-upgrade/cli/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,831 @@
...
+                .value_name("URL")
+                .takes_value(true)
+                .global(true)
+                .validator(|s| is_url_or_moniker(s))
+                .help("JSON RPC URL for the cluster [default: value from configuration file]"),
+        )
+        .subcommand(
+            Command::new("create-escrow").about("Create token account for the program escrow")
+            .arg(
+                Arg::new("original_mint")
```
Comment (CriesofCarrots) [inline token-upgrade/cli/src/main.rs:253]: `.required()`?

File: token-upgrade/cli/src/main.rs [hunk 8]
```diff
@@ -0,0 +1,831 @@
...
+                    .validator(|s| is_valid_pubkey(s))
+                    .value_name("ADDRESS")
+                    .takes_value(true)
+                    .index(1)
+                    .help("Original mint address, whose tokens will be burned")
+            )
+            .arg(
+                Arg::new("new_mint")
```
Comment (CriesofCarrots) [inline token-upgrade/cli/src/main.rs:261]: `.required()`?

File: token-upgrade/cli/src/main.rs [hunk 9]
```diff
@@ -0,0 +1,831 @@
...
+                    .validator(|s| is_valid_signer(s))
+                    .takes_value(true)
+                    .index(3)
+                    .help("Specify the account keypair. This may be a keypair file or the ASK keyword. [default: associated token account for escrow authority]"),
+            )
+        )
+        .subcommand(
+            Command::new("exchange").about("Exchange original tokens for new tokens")
+            .arg(
+                Arg::new("original_mint")
```
Comment (CriesofCarrots) [inline token-upgrade/cli/src/main.rs:280]: `.required()`?

File: token-upgrade/cli/src/main.rs [hunk 10]
```diff
@@ -0,0 +1,831 @@
...
+                    .validator(|s| is_valid_pubkey(s))
+                    .value_name("ADDRESS")
+                    .takes_value(true)
+                    .index(1)
+                    .help("Original mint address, whose tokens will be burned")
+            )
+            .arg(
+                Arg::new("new_mint")
```
Comment (CriesofCarrots) [inline token-upgrade/cli/src/main.rs:288]: `.required()`?

File: token-upgrade/cli/src/main.rs [hunk 11]
```diff
@@ -0,0 +1,831 @@
...
+    }
+    let rpc_client = Arc::new(RpcClient::new_with_commitment(
+        config.json_rpc_url.clone(),
+        config.commitment_config,
+    ));
+
+    match (command, matches) {
+        ("create-escrow", arg_matches) => {
+            let original_mint = pubkey_of(arg_matches, "original_mint").unwrap();
+            let new_mint = pubkey_of(arg_matches, "new_mint").unwrap();
```
Comment (CriesofCarrots) [inline token-upgrade/cli/src/main.rs:392]: See above comments. These args ought to be required, if you're going to unwrap them.

**Review by CriesofCarrots — APPROVED** [review]
r+ required arguments marked

File: token-upgrade/cli/src/main.rs (hunk 11 above)
Comment (joncinque) [reply token-upgrade/cli/src/main.rs:392]: Yep, makes sense, thanks!

**Pull request merged**